controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -0,0 +1,488 @@
+		// not exist."
+		targetPort = getTargetPort(svc, srcPort)
+		exists = true
+	}
+
+	port := &portPublisher{
+		listeners:  []EndpointUpdateListener{},
+		targetPort: targetPort,
+		exists:     exists,
+		k8sAPI:     sp.k8sAPI,
+		log:        sp.log.WithField("port", srcPort),
+	}
+
+	endpoints, err := sp.k8sAPI.Endpoint().Lister().Endpoints(sp.id.Namespace).Get(sp.id.Name)
+	if err == nil {
+		port.updateEndpoints(endpoints)
+	}
+
+	return port
+}
+
+/////////////////////
+/// portPublisher ///
+/////////////////////
+
+func (pp *portPublisher) updateEndpoints(endpoints *corev1.Endpoints) {
+	newPods := pp.endpointsToAddresses(endpoints)
+	if len(newPods) == 0 {
+		pp.exists = true
+		for _, listener := range pp.listeners {
+			listener.NoEndpoints(true)
+		}
+	} else {
+		add, remove := diffPods(pp.pods, newPods)
+		pp.exists = true
+		pp.pods = newPods
+		for _, listener := range pp.listeners {
+			if len(remove) > 0 {
+				listener.Remove(remove)
+			}
+			if len(add) > 0 {
+				listener.Add(add)
+			}
+		}
+	}
+}
+
+func (pp *portPublisher) endpointsToAddresses(endpoints *corev1.Endpoints) PodSet {
+	pods := make(PodSet)
+	for _, subset := range endpoints.Subsets {
+		resolvedPort := pp.resolveTargetPort(subset)
+		for _, endpoint := range subset.Addresses {
+			if endpoint.TargetRef.Kind == "Pod" {
+				id := PodID{
+					Name:      endpoint.TargetRef.Name,
+					Namespace: endpoint.TargetRef.Namespace,
+				}
+				pod, err := pp.k8sAPI.Pod().Lister().Pods(id.Namespace).Get(id.Name)
+				if err != nil {
+					pp.log.Errorf("Unable to fetch pod %v: %s", id, err)
+					continue
+				}
+				ownerKind, ownerName := pp.k8sAPI.GetOwnerKindAndName(pod)
+				pods[id] = Address{
+					Ip:        endpoint.IP,
+					Port:      resolvedPort,
+					Pod:       pod,
+					OwnerName: ownerName,
+					OwnerKind: ownerKind,
+				}
+			}
+		}
+	}
+	return pods
+}
+
+func (pp *portPublisher) resolveTargetPort(subset corev1.EndpointSubset) Port {
+	switch pp.targetPort.Type {
+	case intstr.Int:
+		return Port(pp.targetPort.IntVal)
+	case intstr.String:
+		for _, p := range subset.Ports {
+			if p.Name == pp.targetPort.StrVal {
+				return Port(p.Port)
+			}
+		}
+	}
+	return Port(0)
+}
+
+func (pp *portPublisher) updatePort(targetPort namedPort) {
+	pp.targetPort = targetPort
+	endpoints, err := pp.k8sAPI.Endpoint().Lister().Endpoints(pp.id.Namespace).Get(pp.id.Name)
+	if err == nil {
+		pp.updateEndpoints(endpoints)
+	} else {
+		pp.log.Errorf("Unable to get endpoints during port update: %s", err)
+	}
+}
+
+func (pp *portPublisher) noEndpoints(exists bool) {
+	pp.exists = exists
+	for _, listener := range pp.listeners {
+		listener.NoEndpoints(exists)
+	}
+}
+
+func (pp *portPublisher) subscribe(listener EndpointUpdateListener) {
+	if pp.exists {
+		if len(pp.pods) > 0 {
+			listener.Add(pp.pods)
+		} else {
+			listener.NoEndpoints(true)
+		}
+	} else {
+		listener.NoEndpoints(false)
+	}
+	pp.listeners = append(pp.listeners, listener)
```

Comment:
related to comment on previous review around mutexes (https://github.com/linkerd/linkerd2/pull/2786#discussion_r280855725), if we're guarding access to `portPublisher`'s from a higher-level struct, mind adding comments around which `portPublisher` methods are not thread safe? i'm concerned about future changes that may call these methods and be unaware of that context (for example, when grabbing state for a debug endpoints request).